### ISSUE TITLE
aliasByNode does not work properly when the metric path contains Unicode characters

### ISSUE THREAD

**Issue by zasca** [graphite-project/graphite-web#2024]
affected versions: 0.9.15, 1.0.2 
```
$ find /var/lib/graphite/whisper/app
/var/lib/graphite/whisper/app/
/var/lib/graphite/whisper/app/пример_рф
/var/lib/graphite/whisper/app/пример_рф/m1_rate.wsp
/var/lib/graphite/whisper/app/пример_рф/m15_rate.wsp
/var/lib/graphite/whisper/app/example_com
/var/lib/graphite/whisper/app/example_com/m1_rate.wsp
/var/lib/graphite/whisper/app/example_com/m15_rate.wsp
```

Witnessed Result
```
$ curl -s '127.0.0.1:8080/render/?from=-2hours&target=aliasByNode(app.*.m15_rate,0,-1)&format=json' | jq . | grep targe
    "target": "app.m15_rate"
    "target": "app.app"

$ curl -s '127.0.0.1:8080/render/?from=-2hours&target=aliasByNode(app.*.m15_rate,1,-1)&format=json'
  ...
  File &quot;/opt/graphite/webapp/graphite/render/functions.py&quot;, line 1904, in &lt;genexpr&gt;
    series.name = &#39;.&#39;.join(metric_pieces[n] for n in nodes)
IndexError: list index out of range
```

Expected Result
```
$ curl -s '127.0.0.1:8080/render/?from=-2hours&target=aliasByNode(app.*.m15_rate,0,-1)&format=json' | jq . | grep targe
    "target": "app.m15_rate"
    "target": "app.m15_rate"

$ curl -s '127.0.0.1:8080/render/?from=-2hours&target=aliasByNode(app.*.m15_rate,1,-1)&format=json'
jq . | grep targe
    "target": "example_com.m15_rate"
    "target": "пример_рф.m15_rate"
```

**Comment by zasca** [comment]
looks like the issue is [here](https://github.com/graphite-project/graphite-web/blob/400057dd6e3b8aa0d7116c4f2b82175236219da2/webapp/graphite/render/functions.py#L1861)

**Comment by deniszh** [comment]
Just want to note that Graphite officially doesn't support unicode metrics, only ASCII - https://github.com/graphite-project/graphite-web/issues/243
But your efforts are appreciated, of course.
And, you're right, grammar.py is completely not ready for utf-8 - https://github.com/graphite-project/graphite-web/blob/400057dd6e3b8aa0d7116c4f2b82175236219da2/webapp/graphite/render/grammar.py

**Comment by zasca** [comment]
Thank you, Denis, for that info.

**Comment by piotr1212** [comment]
I've looked into making unicode work a while back. There happens a lot of unicode to str casting which it shouldn't. This should be not that hard to fix. 

The harder part I run into is the font rendering in [render/glyph.py](https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/render/glyph.py). The current implementation uses "Cairo's toy text API" which should be sufficient for most users of latin and cyrillic fonts. The toy API cannot handle more complex situations, for example when a font does not contain a specific glyph it cannot fallback to a different font which does have it. It doesn't support fonts with different writing directions like Chinese and Arabic and doesn't support variable width fonts like Indic and Thai.

The solution would be to implement Pango font rendering. I've did an attempt at this and got it basically working but run into many spacing and positioning issues. I think glyph.py would need some mayor refactoring to get it working properly.
